federationtypes.go

```diff
@@ -590,19 +596,48 @@ type RespProfile struct {
 	AvatarURL   string `json:"avatar_url,omitempty"`
 }
 
-func checkAllowedByAuthEvents(event Event, eventsByID map[string]*Event) error {
+func checkAllowedByAuthEvents(event Event, eventsByID map[string]*Event, missingAuth MissingAuthEventHandler) error {
 	authEvents := NewAuthEvents(nil)
 
 	for _, ae := range event.AuthEventIDs() {
+	retryEvent:
 		authEvent, ok := eventsByID[ae]
 		if !ok {
+			// We don't have an entry in the eventsByID map - neither an event nor nil.
+			if missingAuth != nil {
+				// If we have a MissingAuthEventHandler then ask it for the missing event.
+				if ev, err := missingAuth(ae, event.roomVersion); err == nil {
+					// It claims to have returned an event - populate the eventsByID
+					// map so that we can retry with the new event.
+					eventsByID[ae] = ev
+				} else {
+					// It claims to have not returned an event - put a nil into the
+					// eventsByID map instead. This signals that we tried to retrieve
+					// the event but failed, so we don't keep retrying.
+					eventsByID[ae] = nil
```

Comment:
This is a weird API shape. We don't use the `error` so just make the signature `missingAuth(event, ver) *Event` and always dump it into the map, they can return `nil` if they can't find it.